waiter/src/waiter/token.clj

```diff
@@ -719,3 +760,128 @@
+                                   (async/>! tokens-event-chan))
+                              (assoc current-state :token->token-index (dissoc token->token-index token))))))
+
+                      tokens-watch-channels-update-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "token-watch-maintainer" "process-tokens-channels-update")
+                        (let [{:keys [channel-event watch-chan]} msg
+                              next-state (case channel-event
+                                           :add (do
+                                                  (async/put! watch-chan
+                                                              (make-index-event :INITIAL
+                                                                                (or (vals token->token-index) '())))
+                                                  (async/tap tokens-event-mult watch-chan)
+                                                  (update current-state :watch-count inc))
+                                           :remove (update current-state :watch-count dec)
+                                           (throw (ex-info "Invalid tokens-watch-channels-update-chan event" {:event msg})))]
+                          (log/info "token-watch-maintainer watch count changed!" {:previous-count watches-count
+                                                                                    :new-count (get next-state :watch-count)})
+                          next-state))
+
+                      watch-refresh-timer-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "token-watch-maintainer" "refresh")
+                        (let [next-token->index (get-token->token-index kv-store :refresh true)
+                              [only-old-indexes only-next-indexes _] (data/diff token->token-index next-token->index)
+                              ; if token in old-indexes and not in only-next-indexes, then those token indexes were deleted
+                              delete-events
+                              (for [[token {:keys [owner]}] only-old-indexes
+                                    :when (not (contains? only-next-indexes token))]
+                                (make-index-event :DELETE {:owner owner :token token}))
+                              ; if token in only-next-indexes, it has been updated (soft-delete, and creation included)
+                              update-events
+                              (for [[token _] only-next-indexes]
+                                (make-index-event :UPDATE (get next-token->index token)))
+                              events (concat delete-events update-events)]
+                          (if (empty? events)
+                            (log/info "token-watch-maintainer found no differences in kv-store and current-state")
+                            (do
```

Comment:
we should add a counter and meter to be able to check how often this even finds differences.
It gives us signals to what events might be missed.